File: javascript/src/state/context.js
/** Holds the current state and delegates state-specific requests to it. */
export class Context {
  #state;

  constructor(initial) {
    this.#state = initial;
  }

  request() {
    const before = this.#state.name();
    this.#state.handle(this);
    return `request() handled by ${before}, now in ${this.#state.name()}`;
  }

  setState(state) {
    this.#state = state;
  }

  getStateName() {
    return this.#state.name();
  }
}

File: python/src/patterns/state/context.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .state import State


class Context:
    """Holds the current state and delegates state-specific requests to it."""

    def __init__(self, initial: State) -> None:
        self._state = initial

    def request(self) -> str:
        before = self._state.name()
        self._state.handle(self)
        return f"request() handled by {before}, now in {self._state.name()}"

    def set_state(self, state: State) -> None:
        self._state = state

    def get_state_name(self) -> str:
        return self._state.name()
